runtime/lua/vim/diagnostic.lua

```diff
@@ -827,9 +829,32 @@ local function set_list(loclist, opts)
   local diagnostics = get_diagnostics(bufnr, opts --[[@as vim.diagnostic.GetOpts]], false)
   local items = M.toqflist(diagnostics)
   if loclist then
-    vim.fn.setloclist(winnr, {}, ' ', { title = title, items = items })
+    vim.fn.setloclist(winnr, {}, 'u', { title = title, items = items })
   else
-    vim.fn.setqflist({}, ' ', { title = title, items = items })
+    -- Check if the diagnostics quickfix list no longer exists.
+    if M._qf_id and vim.fn.getqflist({ id = M._qf_id }).id == 0 then
+      M._qf_id = nil
+    end
+
+    -- If we already have a diagnostics quickfix, update it rather than creating a new one.
+    -- This avoids polluting the finite set of quickfix lists, and preserves the currently selected
+    -- entry.
+    if M._qf_id then
+      vim.fn.setqflist({}, 'u', {
+        title = title,
+        items = items,
+        id = M._qf_id,
+      })
+    else
+      vim.fn.setqflist({}, ' ', {
+        title = title,
+        items = items,
+        id = M._qf_id,
+      })
+
+      -- Get the id of the newly created quickfix list.
+      M._qf_id = vim.fn.getqflist({ id = 0 }).id
+    end
```

Comment:
```suggestion
    vim.fn.setqflist({}, M._qf_idwand 'u' or ' ', {
      title = title,
      items = items,
      id = M._qf_id,
    })
    M._qf_id = vim.fn.getqflist({ id = 0 }).id
```

Reply:
Done! Note: this does mean we re-query for the quickfix list id even when we're just updating it. I don't see a problem with that.

(Also, my old code was kind of weird that it passed a `nil` value for `id` when creating a new quickfix list. Thanks to your suggestion, that's gone now.)